### PR TITLE
Fast Inference with vLLM for VLMs

### PR DESCRIPTION
**Pull request by Datta0**
Needed for: unslothai/unsloth#2975

TODO: (Mostly in a follow up PR later)
- [ ] Improve memory calculation for vLLM especially for VLMs
- [ ] Maybe use a logger instead of prints to make the whole process cleaner

### CODE REVIEW

File: unsloth_zoo/empty_model.py [hunk 1]
```diff
@@ -0,0 +1,664 @@
...
+            setattr(config_obj, attr, value)
+pass
+
+
+@torch.inference_mode()
+def create_empty_vision_model(config, dtype = torch.float16):
+    # All Unsloth Zoo code licensed under LGPLv3
+    model_type = config.model_type
+
+    from transformers.models.siglip.modeling_siglip import SiglipVisionModel
```
Comment (danielhanchen) [inline unsloth_zoo/empty_model.py:262]: Is it always Siglip? Does vLLM support anything else?

File: unsloth_zoo/empty_model.py [hunk 2]
```diff
@@ -0,0 +1,664 @@
...
+    else:
+        language_model_prefix = "model"
+        language_model = new_model.model
+
+    # Embeddings
+    embed_tokens_key = f"{language_model_prefix}.embed_tokens.weight"
+    language_model.embed_tokens = torch.nn.Embedding.from_pretrained(
+        quant_state_dict[embed_tokens_key],
+        freeze = True,
+        padding_idx = config.pad_token_id,
```
Comment (danielhanchen) [inline unsloth_zoo/empty_model.py]: Check if padding idx goes out of bounds (I think it's `quant_state_dict[embed_tokens_key].shape[0] -1` or was it `shape[1]`

File: unsloth_zoo/vllm_utils.py [hunk 3]
```diff
@@ -47,12 +47,14 @@
...
 from functools import partial
 from .utils import _get_dtype
+from .empty_model import *
 from .patching_utils import patch_model_and_tokenizer
 from .temporary_patches.common import (
     get_torch_compile_options,
     UNSLOTH_ENABLE_LOGGING,
 )
-from unsloth import DEVICE_TYPE
+# from unsloth import DEVICE_TYPE
```
Comment (danielhanchen) [inline unsloth_zoo/vllm_utils.py]: Remove change

File: unsloth_zoo/vllm_utils.py [hunk 4]
```diff
@@ -745,138 +749,206 @@ def vllm_dynamic_quant_supported(
...
-            weights = llm.collective_rpc("get_weight_ipc_handles", args = tuple())[0]
-            weights = weights[gpu_ids[0]]
-            for weight_name, (to_cuda_fx, cuda_data,) in weights.items():
-                vllm_state_dict[weight_name] = to_cuda_fx(*cuda_data)
-            pass
-            raise NotImplementedError("Unsloth: Currently vLLM RPC is not yet fully enabled!")
-        except Exception as e:
-            raise RuntimeError(f"Unsloth: Cannot get internal vLLM states with error = {str(e)}")
+    except Exception as e:
+        # If we can't access the model directly, raise a more informative error
```
Comment (danielhanchen) [inline unsloth_zoo/vllm_utils.py]: Do not delete the code we have, just leave it and raise the error

File: unsloth_zoo/vllm_utils.py [hunk 5]
```diff
@@ -745,138 +749,206 @@ def vllm_dynamic_quant_supported(
...
+        raise RuntimeError(f"Unsloth: Cannot access vLLM internal model. This might be due to a vLLM version incompatibility. Error: {str(e)}")
     pass
 
+    print(f"Unsloth: vllm_internals: \n\n{vllm_internals}\n\n")
```
Comment (danielhanchen) [inline unsloth_zoo/vllm_utils.py]: Here

File: unsloth_zoo/vllm_utils.py [hunk 6]
```diff
@@ -745,138 +749,206 @@ def vllm_dynamic_quant_supported(
...
-                state_dict[prefix + ".weight." + k] = v
-            pass
-        else:
-            # Normal FP16 weights
-            qweight.requires_grad_(False) # Disable grad - sometimes vLLM forgets
-            state_dict[prefix + ".weight"] = qweight[dim_offsets[kk] : dim_offsets[kk + 1]]
-            quant_state_dict[prefix + ".weight"] = state_dict[prefix + ".weight"]
-        pass
+    def get_state_dict(prefix, kk, state_dict, proj, slice_weights=True, slice_index=-1):
+        try:
```
Comment (danielhanchen) [inline unsloth_zoo/vllm_utils.py]: Do not do try except

File: unsloth_zoo/vllm_utils.py [hunk 7]
```diff
@@ -745,138 +749,206 @@ def vllm_dynamic_quant_supported(
...
+            # https://github.com/huggingface/transformers/blob/1cea763ba422b83778a8db0374ea90f43b09992b/src/transformers/models/mllama/modeling_mllama.py#L1147
+            shrink_size = getattr(proj,"org_vocab_size", vocab_size)
+            if shrink_size and ("embed_tokens" in prefix or "lm_head" in prefix):
+                if weight.shape[0] > shrink_size:
+                    weight = weight[:shrink_size]
+
+            state_dict[prefix + ".weight"] = weight
+            quant_state_dict[prefix + ".weight"] = weight
+
+            # Handle bias
```
Comment (danielhanchen) [inline unsloth_zoo/vllm_utils.py]: Confirm if removing indentation due to try except that logic is equivalent as before

File: unsloth_zoo/vllm_utils.py [hunk 8]
```diff
@@ -1265,19 +1328,28 @@ def load_vllm(
         account_for_gradients = training,
     )
 
-    # Check max_num_batched_tokens for max_seq_length
-    # Must be >= max_num_batched_tokens
-    if max_num_batched_tokens <= 0:
-        max_seq_length = 256
-        max_num_batched_tokens = 256
+    enable_chunked_prefill = True
```
Comment (danielhanchen) [inline unsloth_zoo/vllm_utils.py:1405]: Are you sure chunked prefill works for Tesla T4 Colab LoRA / QLoRA?

Comment (danielhanchen) [reply unsloth_zoo/vllm_utils.py:1405]: Does vllm 0.10.0 work now? How about 0.8.5post1

File: unsloth_zoo/vllm_utils.py [hunk 9]
```diff
@@ -1467,19 +1549,25 @@ def load_vllm(
 
         disable_log_stats      = disable_log_stats,
         enable_prefix_caching  = enable_prefix_caching,
-        # enable_chunked_prefill = True, # LoRA fails with chunked prefill as at Feb 2025
+        enable_chunked_prefill = enable_chunked_prefill, # LoRA fails with chunked prefill as at Feb 2025
```
Comment (danielhanchen) [inline unsloth_zoo/vllm_utils.py:1628]: Confirm chunked prefill actually works with LoRA / QLoRA

File: unsloth_zoo/vllm_utils.py [hunk 10]
```diff
@@ -1467,19 +1549,25 @@ def load_vllm(
...
         # max_seq_len_to_capture fails for V1
         # max_seq_len_to_capture = min(8192, max_seq_length + 256), # Default is 8192 for CUDAGraphs
         compilation_config     = compilation_config, # 0, 1, 2, 3
-        enforce_eager          = enforce_eager,
+        enforce_eager          = True,
```
Comment (danielhanchen) [inline unsloth_zoo/vllm_utils.py]: Here

File: unsloth_zoo/vllm_utils.py [hunk 11]
```diff
@@ -2024,105 +2150,132 @@ def _test_get_vllm_state_dict(
...
+            import transformers
+            model_class = getattr(transformers, config.architectures[0])
+        else:
+            raise ValueError(f"Unsloth: Model type {model_type} not supported for vision models")
+
+    print(f'Loading model with type {model_class}')
+    model = model_class.from_pretrained(
         model_name,
-        device_map          = "sequential",
+        device_map          = "auto",
```
Comment (danielhanchen) [inline unsloth_zoo/vllm_utils.py]: WHy auto

File: unsloth_zoo/vllm_utils.py [hunk 12]
```diff
@@ -2182,19 +2335,3 @@ def test_get_vllm_state_dict():
         torch.cuda.empty_cache()
     pass
 pass
-
```
Comment (danielhanchen) [inline unsloth_zoo/vllm_utils.py:2243]: Add back

File: unsloth_zoo/empty_model.py (hunk 1 above)
Comment (Datta0) [reply unsloth_zoo/empty_model.py:262]: Among the models we support so far, only SigLIP seems to be causing issues. But if we notice any other model causing issues in the future, we can repeat something similar

File: unsloth_zoo/vllm_utils.py (hunk 8 above)
Comment (Datta0) [reply unsloth_zoo/vllm_utils.py:1405]: If you see the [notebook here](https://colab.research.google.com/drive/1JAxPxxQmxW5Dkwno9wcpZd9xEc-4MNTw?usp=sharing) there's a line in model load cell
`INFO 07-28 08:03:08 [llm_engine.py:228] Initializing a V0 LLM engine (v0.10.0) with config: model='unsloth/qwen2.5-vl-7b-instruct-unsloth-bnb-4bit', speculative_config=None, tokenizer='unsloth/qwen2.5-vl-7b-instruct-unsloth-bnb-4bit', ...` (truncated)

File: unsloth_zoo/empty_model.py [hunk 13]
```diff
@@ -0,0 +1,678 @@
+__all__ = [
```
Comment (danielhanchen) [inline unsloth_zoo/empty_model.py:17]: License header

File: unsloth_zoo/empty_model.py [hunk 14]
```diff
@@ -0,0 +1,678 @@
...
+            "model.vision_tower.vision_model.embeddings.patch_embedding",
+            "model.vision_tower.vision_model.embeddings.position_embedding",
+            "model.vision_tower.vision_model.post_layernorm",
+            "model.multi_modal_projector.mm_input_projection_weight",
+            "model.vision_model.post_tile_positional_embedding.gate",
+            "model.vision_model.gated_positional_embedding.tile_embedding",
+            "model.vision_model.pre_tile_positional_embedding.embedding",
+            "model.vision_model.gated_positional_embedding",
+            "model.vision_model.post_tile_positional_embedding.embedding",
+            "model.vision_model.pre_tile_positional_embedding.gate"
```
Comment (danielhanchen) [inline unsloth_zoo/empty_model.py:564]: Very good work

File: unsloth_zoo/vllm_utils.py [hunk 15]
```diff
@@ -494,6 +495,7 @@ def patch_vllm_enable_sleep_mode():
     from typing import Optional, Union, Tuple
 
     logger = init_logger(__name__)
+    print(f"Unsloth: Enabling vLLM standby mode")
```
Comment (danielhanchen) [inline unsloth_zoo/vllm_utils.py]: Change to logger.info

File: unsloth_zoo/vllm_utils.py [hunk 16]
```diff
@@ -717,6 +719,7 @@ def capture_model_wrapper_v0(self, *args, **kwargs):
 def patch_vllm(debug = True):
     # Temporary patch to disable multiprocessing for vLLM
     # Allows accessing model_executor
+    print(f'Unsloth: Patching vLLM')
```
Comment (danielhanchen) [inline unsloth_zoo/vllm_utils.py]: Change to logger info

File: unsloth_zoo/vllm_utils.py (hunk 9 above)
Comment (danielhanchen) [reply unsloth_zoo/vllm_utils.py:1628]: Confirm please!

Comment (Datta0) [reply unsloth_zoo/vllm_utils.py:1628]: Oh this does work without any problems in my testing

Comment (danielhanchen) [reply unsloth_zoo/vllm_utils.py:1628]: Maybe it's vLLM versions? Best to add a check if vllm version is less than 0.10.0 or something, delete this kwarg

File: unsloth_zoo/vllm_utils.py [hunk 17]
```diff
@@ -490,12 +492,33 @@ def unpatch_bitsandbytes_compute_dtype():
...
-    from vllm.device_allocator.cumem import CuMemAllocator, libcudart, unmap_and_release, create_and_map
-    from vllm.logger import init_logger
+    from vllm.device_allocator.cumem import CuMemAllocator, libcudart, unmap_and_release, create_and_map, AllocationData
     from vllm.utils import is_pin_memory_available
-    from typing import Optional, Union, Tuple
+    from typing import Optional, Union, Tuple, Any
+
+    logger.info(f"Unsloth: Enabling vLLM standby mode")
+
+    def __init__(self):
```
Comment (danielhanchen) [inline unsloth_zoo/vllm_utils.py:501]: Here
